packages/flutter_tools/lib/src/flutter_project_metadata.dart

```diff
@@ -0,0 +1,99 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:yaml/yaml.dart';
+
+import 'base/file_system.dart';
+import 'base/logger.dart';
+import 'base/utils.dart';
+
+enum FlutterProjectType {
+  /// This is the default project with the user-managed host code.
+  /// It is different than the "module" template in that it exposes and doesn't
+  /// manage the platform code.
+  app,
+  /// The is a project that has managed platform host code. It is an application with
+  /// ephemeral .ios and .android directories that can be updated automatically.
+  module,
+  /// This is a Flutter Dart package project. It doesn't have any native
+  /// components, only Dart.
+  package,
+  /// This is a native plugin project.
+  plugin,
+}
+
+extension FlutterProjectTypeExtension on FlutterProjectType {
+  String get name => getEnumName(this);
+}
+
+FlutterProjectType stringToProjectType(String value) {
+  FlutterProjectType result;
+  for (final FlutterProjectType type in FlutterProjectType.values) {
+    if (value == type.name) {
+      result = type;
+      break;
+    }
+  }
+  return result;
+}
+
+/// A wrapper around the `.metadata` file.
+class FlutterProjectMetadata {
+  FlutterProjectMetadata(
+    File metadataFile,
+    Logger logger,
+    )  : _metadataFile = metadataFile,
+      _logger = logger;
+
+  final File _metadataFile;
+  final Logger _logger;
+
+  String get versionChannel => _versionValue('channel');
+  String get versionRevision => _versionValue('revision');
+
+  FlutterProjectType get projectType {
+    final dynamic projectTypeYaml = _metadataValue('project_type');
+    if (projectTypeYaml is String) {
+      return stringToProjectType(projectTypeYaml);
+    } else {
+      _logger.printTrace('.metadata project_type version is malformed.');
```

Comment:
These used to be `throwToolExit`, what will happen now?

Reply:
Recreating the project with missing metadata won't work. create_tests that fail:
- `Will create an app project if non-empty non-project directory exists without .metadata`
- `cannot create a project if non-empty non-project directory exists with .metadata` will throw the wrong toolExit.

Reply:
What will happen if someone corrupts their .metadata file? Will the tool throw a ToolExit or send a report to crash logging?

Reply:
Let me move the throws back into the create command.  We don't want it to throw when it's generating a crash report.

Reply:
#### Before this change getting the project type
If the .metadata is present but not a Yaml map:
- throws "pubspec.yaml is malformed." (yes, pubspec, not metadata).

If project_type isn't a String:
- throws ".metadata is malformed."

#### After this change
If .metadata is malformed, either not a Yaml map or project_type isn't a String:
   - Checks for the presence of directories to tell only if it's an app.  If it's an app, carry on with template reconstruction
   - If those directories are missing, throw "'Sorry, unable to detect the type of project to recreate. Try creating a fresh project and migrating your existing code to the new project manually."

I think the second behavior is actually better:
1. It doesn't have the misleading "pubspec" message
2. ".metadata is malformed" is a pretty lame message and doesn't tell the user what to do.
3. If it can figure out the project type, it does.
4. If it can't, it gives an actually useful recovery message.

What do you think?

Reply:
That sounds good to me.

Reply:
Is the "Sorry, unable to detect..." error message in this PR, or is it pre-existing?

Reply:
Sorry should have noted, it's pre-existing.
https://github.com/flutter/flutter/blob/685e9d1e47d80bd3bd22d97e1774fa264930434c/packages/flutter_tools/lib/src/commands/create.dart#L294

Updated:
#### Before this change getting the project type
If the .metadata is present but not a Yaml map:
- throws "pubspec.yaml is malformed." (yes, pubspec, not metadata).

If project_type isn't a String:
- throws ".metadata is malformed."

If the .metadata isn't present and the directories don't indicate it's an app:
- throw "'Sorry, unable to detect the type of project to recreate. Try creating a fresh project and migrating your existing code to the new project manually."